### PR TITLE
Web Story Embed block for Gutenberg

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

This PR adds a new `web-stories/embed` block to be used in Gutenberg. This new block can be used to display stories from the own site and other sites using `amp-story-player`.

## Relevant Technical Choices

* Dynamic block, makes it easier to only load the scripts & styles when needed. Makes future changes easier, too.
* `save.js` is still used, as a fallback/cache should be plugin ever be disabled.
* Styling is similar to regular embed blocks in Gutenberg, for better look and feel.
* New REST API endpoint that returns the story title and poster for a given URL. That data is needed for the player.
* The REST API controller short-circuits for URLs from the own site in order to not hit the own site with unnecessary HTTP requests.

### To-do

**Editor**

* [x] WYSIWYG preview
* [x] REST API proxy endpoint
* [x] When entering a URL, fetch title and poster image via REST API proxy endpoint
* [x] Ability to resize the player using slider controls
* [x] Ability to override poster via sidebar
* [x] Ability to resize the player using input fields in sidebar
* [x] Make title editable in sidebar
* [x] Fix loading issue with `amp-story-player` (https://github.com/ampproject/amphtml/issues/27331?)
  **Suggestion:** Implement in follow-up PR.
* [ ] Dropdown to select posts from the stories editor?  
  **Suggestion:** Implement in follow-up PR.

**Frontend**

* [x] Implement `render_callback`
* [ ] Display Fullscreen toggle within player to navigate to story  
  Got to URL? Or fullscreen iframe with history update?  
  **Suggestion:** Implement in follow-up PR.

**Code**

* [ ] **UX:** Provide custom block icon
* [ ] **UX / Product:** Define better example story
* [ ] Tests
  * [x] Fix `WP_Block_Type_Registry::register` notices in tests
  * [ ] E2E tests once stable  
      **Suggestion:** Implement in follow-up PR.
  * [ ] Unit tests for individual components  
    There are currently unresolved issues with mocking `@wordpress/element` that need to be figured out.
    Can perhaps use https://jestjs.io/docs/en/jest-object#jestgenmockfrommodulemodulename to mock `@wordpress/element` & co. See also https://github.com/ampproject/amp-wp/blob/1.4/tests/shared/test-utils/wp-element-mock.js
      **Suggestion:** Implement in follow-up PR.

## Demo

Click below to expand:

<details>
<summary>Screenshots</summary>

Screenshot 2020-04-16 at 16 41 28
Screenshot 2020-04-29 at 13 54 19
Screenshot 2020-05-04 at 16 56 52
Screenshot 2020-05-04 at 16 56 19
</details>

----

See #1118.

### CODE REVIEW

**Comment by swissspidy** [comment]
Things I have tried for the in-editor preview so far:

* Using the provided [`Sandbox`](https://github.com/WordPress/gutenberg/tree/0504dad2685682c868c0ce7c54c71172063b5aa6/packages/components/src/sandbox) component which creates an iframe with custom HTML and assets injected. Seemed to work fine except for some issues:
  * [This line](https://github.com/WordPress/gutenberg/blob/0504dad2685682c868c0ce7c54c71172063b5aa6/packages/components/src/sandbox/index.js#L133) causes CORS issues when loading the story player CSS
  * Due to being in an iframe, the player complained about the window being too small. 
* Directly adding `amp-story-player` in markup  
  The player was stuck in loading state forever

**Comment by swissspidy** [comment]
@spacedmonkey @miina would be nice to get your eyes on this already, even though it's still WIP.

**Comment by spacedmonkey** [comment]
I will take a look

File: includes/Embed_Block.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	/**
+	 * Renders the block type output for given attributes.
+	 *
+	 * @param array  $attributes Block attributes.
+	 * @param string $content    Block content.
+	 *
+	 * @return string Rendered block type output.
+	 */
+	public function render_block( array $attributes, $content ) {
+		return $content;
```
Comment (spacedmonkey) [inline includes/Embed_Block.php]: What does this do?

Comment (swissspidy) [reply includes/Embed_Block.php]: Nothing... yet. Will probably remove if not needed.

File: assets/src/story-embed-block/block/save.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+function save({ attributes }) {
+  const { url, title, poster, width, height, align = 'none' } = attributes;
+
+  if (!url || !title) {
+    return null;
+  }
+
+  return (
+    <div className={`wp-block-web-stories-embed align${align}`}>
```
Comment (miina) [inline assets/src/story-embed-block/block/save.js:49]: The original plan was to create a dynamic block and not save content directly into the post, curious why the change? I guess currently the script and style are always loaded now, even if the block is not used in a post?

File: assets/src/story-embed-block/block/block.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "web-stories/embed",
+  "category": "embed",
+  "attributes": {
```
Comment (miina) [inline assets/src/story-embed-block/block/block.json:4]: If it's not a dynamic block then these attributes could use a source instead of being saved into comments.

File: assets/src/story-embed-block/block/save.js (hunk 2 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/save.js:49]: > The original plan was to create a dynamic block and not save content directly into the post, curious why the change?

I might change it again to a dynamic block, haven't decided yet. If we keep the output that simple, there wouldn't really be a need for a dynamic block.

> I guess currently the script and style are always loaded now, even if the block is not used in a post?

Yeah that's a limitation of how `register_block_type` registers blocks. I'd change that with a dynamic block.

File: assets/src/story-embed-block/block/edit.js [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+      setCannotEmbed(true);
+    }
+  }, [outerURL, poster, title]);
+
+  useEffect(() => {
+    if (!storyData) {
+      return;
+    }
+
+    if (!storyData.title) {
```
Comment (miina) [inline assets/src/story-embed-block/block/edit.js]: Could this be just `if (!storyData?.title)`?

File: assets/src/story-embed-block/block/block.json (hunk 3 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/block.json:4]: Made it a partially dynamic block now, keeping the `save()` function as a cached copy.

File: assets/src/story-embed-block/block/save.js (hunk 2 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/save.js:49]: Made it a partially dynamic block now, keeping the `save()` function as a cached copy.

File: includes/Embed_Block.php (hunk 1 above)
Comment (swissspidy) [reply includes/Embed_Block.php]: Updated now.

**Comment by Enriqe** [comment]
> Fix loading issue with amp-story-player

Can I ask what the issue you are seeing is?

**Comment by swissspidy** [comment]
@enriqe happy to provide a sandbox site / video or connect offline, but in short: the player script is already on the page, but the player component is added later on by the user. Doing so does not properly load the player, instead it keeps showing the poster image with the loading indicator. I assume the upcoming programmatic API might help here?

**Comment by Enriqe** [comment]
Got it, thanks for the context @swissspidy. Yeah, the upcoming programmatic API should solve this use case. As I am currently developing it, it might be a good idea for you to send me a sandbox / demo to make sure I am solving for your use case. Thanks!

File: assets/src/story-embed-block/block/edit.js [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+          setStoryData(data);
+        })
+        .catch(() => {
+          setCannotEmbed(true);
+        })
+        .finally(() => {
+          setIsFetchingData(false);
+        });
+    } catch {
+      setCannotEmbed(true);
```
Comment (miina) [inline assets/src/story-embed-block/block/edit.js]: Is the outer `try`-`catch` necessary here with the `apiFetch` already catching?

File: assets/src/story-embed-block/block/edit.js [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    try {
+      setIsFetchingData(true);
+      const urlObj = new URL(outerURL);
+
+      apiFetch({
+        path: `web-stories/v1/embed?url=${urlObj.toString()}`,
+      })
+        .then((data) => {
+          setCannotEmbed(!data.title);
+          setStoryData(data);
```
Comment (miina) [inline assets/src/story-embed-block/block/edit.js]: Is there a reason to set the story data even if `cannotEmbed` (`!data.title`)?

File: includes/REST_API/Embed_Controller.php [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+	 * Callback for the Web Stories embed endpoints.
+	 *
+	 * Returns information about the given story.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_proxy_item( $request ) {
+		$url = untrailingslashit( $request['url'] );
```
Comment (miina) [inline includes/REST_API/Embed_Controller.php:95]: Logic for if the `url` isn't set?

File: includes/REST_API/Embed_Controller.php [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$title  = '';
+		$poster = '';
+
+		$data = [
+			'title'  => $title,
```
Comment (miina) [inline includes/REST_API/Embed_Controller.php]: Why not assign `''` directly here?

File: includes/REST_API/Embed_Controller.php [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+	/**
+	 * Retrieve content of a given DOM node attribute.
+	 *
+	 * @param DOMNodeList<DOMElement>|false $query XPath query result.
+	 * @param string                        $attribute Attribute name.
+	 *
+	 * @return string|false Attribute content on success, false otherwise.
+	 */
+	private function get_dom_attribute_content( $query, $attribute ) {
```
Comment (miina) [inline includes/REST_API/Embed_Controller.php:282]: Looks like the same as in `Link_Controller`? Could be in a Trait in theory. Not a huge difference currently though.

File: includes/REST_API/Embed_Controller.php [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), [ 'status' => 404 ] );
+		}
+
+		$html = wp_remote_retrieve_body( $response );
+
+		if ( ! $html ) {
+			set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
```
Comment (miina) [inline includes/REST_API/Embed_Controller.php]: Wondering what if the error is temporary? Should we perhaps not save it to cache in case of erroring?

File: includes/REST_API/Embed_Controller.php [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+		$poster    = $this->get_dom_attribute_content( $amp_story, 'poster-portrait-src' );
+
+		$data = [
+			'title'  => $title ?: '',
+			'poster' => $poster ?: '',
+		];
+
+		set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+
+		return rest_ensure_response( $data );
```
Comment (miina) [inline includes/REST_API/Embed_Controller.php]: This should maybe return an error if an actual story wasn't found from the HTML?

File: assets/src/story-embed-block/block/save.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+ * The server-side 'render_callback' is used to override this on page load.
+ *
+ * @see https://developer.wordpress.org/block-editor/developers/block-api/block-edit-save/
+ *
+ * @param {Object} props Props.
+ * @param {Object} props.attributes Block attributes.
+ * @return {null|*} Rendered block.
+ */
+function save({ attributes }) {
+  const { url, title, poster, width, height, align = 'none' } = attributes;
```
Comment (miina) [inline assets/src/story-embed-block/block/save.js:42]: Hmm, does `save` do anything at all if it's a dynamic block rendering the output? As far as I remember it was just ignored in this case.

**Comment by miina** [comment]
@swissspidy From testing:

Currently, when I add the block, it starts loading right away as if an URL would already be added.
Then, when I saved and reloaded, it embedded the wp-admin posts lists instead (`.../wp-admin/edit.php`) without me doing anything. Very odd.

I'm not able to test it currently due to these oddities :) Does it work as expected for you?

**Comment by miina** [comment]
Seeing some console errors upon adding the block:
Screenshot 2020-05-05 at 12 57 00

File: assets/src/story-embed-block/block/edit.js (hunk 5 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/edit.js]: The catch is for the URL constructor throwing on invalid URLS

File: assets/src/story-embed-block/block/block.json [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    "url": {
+      "type": "string"
```
Comment (spacedmonkey) [inline assets/src/story-embed-block/block/block.json]: Should this be type of url?

**Review by spacedmonkey — DISMISSED** [review]
I found this PR, impossible to test, even after a rebase with master. 
When a block is inserted, I see this view.
Screenshot 2020-05-06 at 14 36 36
With this console warnings. 
Screenshot 2020-05-06 at 14 36 56

When I finally get into the embed view and put a local url, it didn't work. 

Screenshot 2020-05-06 at 14 45 28

There seems to a problem with the REST API. 

Screenshot 2020-05-06 at 14 44 51

Until this PR is working, it is hard to test. @swissspidy can you fix and request a new review.

File: includes/REST_API/Embed_Controller.php (hunk 7 above)
Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php:95]: It's a required param, doesn't the REST API already handle those?

File: includes/REST_API/Embed_Controller.php (hunk 10 above)
Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php]: Temporary errors are likely all caught by the status check above, which doesn't set a transient.

File: assets/src/story-embed-block/block/save.js (hunk 12 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/save.js:42]: Yes. As per the doc block above:

> Represents a cached copy of the block’s content to be shown in case the plugin is disabled.
> 
> The server-side 'render_callback' is used to override this on page load.

This use case is mentioned in the handbook: https://developer.wordpress.org/block-editor/developers/block-api/block-edit-save/

**Comment by swissspidy** [comment]
@miina @spacedmonkey I think those two big issues should be resolved now

File: includes/REST_API/Embed_Controller.php (hunk 7 above)
Comment (miina) [reply includes/REST_API/Embed_Controller.php:95]: Ah, I was mainly thinking of an `empty` case here (`''`), that would still pass through.
Although I guess eventually it would return a `'rest_invalid_url'`, however, perhaps an empty URL error would be more accurate. Thoughts?

Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php:95]: Got it. Adding a test case and optimizing that sounds reasonable

File: includes/Embed_Block.php [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+	/**
+	 * Renders the block type output for given attributes.
+	 *
+	 * @param array  $attributes Block attributes.
+	 * @param string $content    Block content.
+	 *
+	 * @return string Rendered block type output.
+	 */
+	public function render_block( array $attributes, $content ) {
+		$defaults = [
```
Comment (miina) [inline includes/Embed_Block.php]: These could also be set in `register_block_type` under `attributes` with types and defaults perhaps?

Comment (swissspidy) [reply includes/Embed_Block.php]: Totally 👍

File: assets/src/story-embed-block/block/save.js (hunk 12 above)
Comment (miina) [reply assets/src/story-embed-block/block/save.js:42]: Nice, thanks for the info! 👍

File: assets/src/story-embed-block/block/storyPlayer.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+function StoryPlayer({ url, title, poster, width, height }) {
+  return (
+    <amp-story-player
+      style={{
+        width: width ? `${width}px` : undefined,
+        height: height ? `${height}px` : undefined,
+      }}
+    >
+      <a
+        href={url}
```
Comment (miina) [inline assets/src/story-embed-block/block/storyPlayer.js:38]: Should this perhaps open in another tab?
Or alternatively, the whole content could be disabled for interaction. Thoughts?

Comment (swissspidy) [reply assets/src/story-embed-block/block/storyPlayer.js:38]: For the editor we have to do something, yeah. But for save/frontend this is the expected markup.

In the editor I think it‘s only an issue when the player hasn‘t loaded yet. After page reload the interaction should work fine, no?

Comment (miina) [reply assets/src/story-embed-block/block/storyPlayer.js:38]: It looks like clicking on the block leaves the editor and opens the embedded story instead if a story is loaded. Probably it shouldn't leave the editor when clicking on a block, e.g. if I click on the block to select it, I'm redirected instead.

Comment (swissspidy) [reply assets/src/story-embed-block/block/storyPlayer.js:38]: The issue is that the player currently doesn't load properly in the editor (noted in the PR description + discussion above). Once it does, it would behave like on the frontend: clicking on the story would just tap throuhg pages, not navigate away.

Comment (miina) [reply assets/src/story-embed-block/block/storyPlayer.js:38]: Oh, okay, now I see what you meant by "loading properly".

Hmm, okay. Yes, perhaps the thing is that it didn't load properly for me when opening an already saved post with embeds (or after reload). Will check again a bit later, didn't focus on that when trying it out.

**Comment by spacedmonkey** [comment]
I did some more testing on this PR. 
I was unable to preview stories on the local dev box. I am getting this error from PHP 
`cURL error 7: Failed to connect to localhost port 8899: Connection refused`. 

When I use a test url from staging https://stories-new-wordpress-amp.pantheonsite.io/stories/auto-draft-6/ I get this. 

Screenshot 2020-05-08 at 16 04 02
See like the styling is broken here. The spinner next goes. 

@swissspidy Can you take a look into this or give me steps on how to test this PR.

**Comment by swissspidy** [comment]
@spacedmonkey the broken preview in the editor is currently expected, see discussion above with Enriqe and open TODO in PR description. This is not a fault in the block, but one in the player.

As for the curl error, it might be an issue with loopback requests in the Docker environment.

File: assets/src/story-embed-block/block/edit.js [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+    align = 'none',
+    poster,
+    title,
+  } = attributes;
+
+  const [editingURL, setEditingURL] = useState(false);
+  const [localURL, setLocalURL] = useState(outerURL);
+  const [isFetchingData, setIsFetchingData] = useState(false);
+  const [storyData, setStoryData] = useState({});
+  const [cannotEmbed, setCannotEmbed] = useState('');
```
Comment (spacedmonkey) [inline assets/src/story-embed-block/block/edit.js]: this default to `false`

File: includes/REST_API/Embed_Controller.php [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+			);
+
+			$qv = [
+				'domain'                 => $url_parts['host'],
+				'path'                   => '/',
+				'update_site_meta_cache' => false,
+			];
+
+			// In case of subdirectory configs, set the path.
+			if ( ! is_subdomain_install() ) {
```
Comment (spacedmonkey) [inline includes/REST_API/Embed_Controller.php:190]: What about multi network?

File: includes/REST_API/Embed_Controller.php [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+				[
+					'host' => '',
+					'path' => '/',
+				]
+			);
+
+			$qv = [
+				'domain'                 => $url_parts['host'],
+				'path'                   => '/',
+				'update_site_meta_cache' => false,
```
Comment (spacedmonkey) [inline includes/REST_API/Embed_Controller.php:186]: Limit the query to 1. Do not prime meta caches

File: includes/REST_API/Embed_Controller.php (hunk 9 above)
Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:282]: Agreed

File: includes/REST_API/Embed_Controller.php (hunk 18 above)
Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:186]:
```
'number'                 => 1,
'update_site_cache'      => false,
'update_site_meta_cache' => false,
```

Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:186]: Also, don't we only care about the site id here?

File: assets/src/story-embed-block/block/edit.js (hunk 16 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/edit.js]: Oh, an oversight from when I tried something else today. Good catch!

File: includes/REST_API/Embed_Controller.php [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+			// In case of subdirectory configs, set the path.
+			if ( ! is_subdomain_install() ) {
+				$path    = explode( '/', ltrim( $url_parts['path'], '/' ) );
+				$path    = reset( $path );
+				$network = get_network();
+				if ( $path && $network instanceof \WP_Network ) {
+					$qv['path'] = $network->path . $path . '/';
+				}
+			}
+
```
Comment (spacedmonkey) [inline includes/REST_API/Embed_Controller.php:198]: I think the query variables needs a filter.

File: includes/REST_API/Embed_Controller.php (hunk 18 above)
Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php:186]: For reference, this is largely based on `get_oembed_response_data_for_url`. Might want to patch that one in core?

File: includes/REST_API/Embed_Controller.php (hunk 17 above)
Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php:190]: For reference, this is largely based on `get_oembed_response_data_for_url`.

If we don't catch all cases it's not a big deal, it will just result in an HTTP request.

File: includes/REST_API/Embed_Controller.php (hunk 19 above)
Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php:198]: Use case? I don't necessarily like adding filters everywhere when there is no clear need for it.

For reference, this is largely based on `get_oembed_response_data_for_url`. 
Would a filter for the post ID like in that function be good enough?

File: includes/REST_API/Embed_Controller.php [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+	 * Returns information about the given story.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_proxy_item( $request ) {
+		$url = untrailingslashit( $request['url'] );
+
+		if ( empty( $url ) ) {
```
Comment (spacedmonkey) [inline includes/REST_API/Embed_Controller.php:97]: Most of this logic seems repeat of logic found in the links controller. Can we use a trait or another class that is extended.

Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php:97]: For reference, same question here from Miina:

https://github.com/google/web-stories-wp/pull/1249#discussion_r420020108

I don't think at this point it's worth optimizing this too much.

In the future this block & endpoint will probably be split up into its own package/plugin anyway, so might be best to keep things decoupled.

File: includes/REST_API/Embed_Controller.php (hunk 18 above)
Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:186]: I will take a look.

File: includes/REST_API/Embed_Controller.php (hunk 19 above)
Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:198]: If you have a very custom multisite site, this query will likely not find anything. A filter would allow a developer to fix any lookups. Not really required, just a nice to have.

**Comment by spacedmonkey** [comment]
> @spacedmonkey the broken preview in the editor is currently expected, see discussion above with Enriqe and open TODO in PR description. This is not a fault in the block, but one in the player.

I don't see the TODO you are referencing here.

File: assets/src/story-embed-block/block/edit.js (hunk 16 above)
Comment (spacedmonkey) [reply assets/src/story-embed-block/block/edit.js]: Screenshot 2020-05-11 at 14 59 32

I think it will fix a warning that I am seeing.

**Comment by swissspidy** [comment]
@spacedmonkey It's the one starting with "Fix loading issue with `amp-story-player` ..."

File: assets/src/story-embed-block/block/index.js [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+// TODO: Replace with web stories icon.
```
Comment (spacedmonkey) [inline assets/src/story-embed-block/block/index.js:21]: Can we fix this now?

Comment (swissspidy) [reply assets/src/story-embed-block/block/index.js:21]: We don't have such an icon / logo yet. Needs UX to provide one for us.

Mentioned as a todo in the PR description:

> UX: Provide custom block icon

Also mentioned on the issue itself:

https://github.com/google/web-stories-wp/issues/1118#issuecomment-624894364

Comment (spacedmonkey) [reply assets/src/story-embed-block/block/index.js:21]: Can we use the story icon we are using for the CPT for now?

**Comment by spacedmonkey** [comment]
> @spacedmonkey It's the one starting with "Fix loading issue with `amp-story-player` ..."

I dont see that in the code base.

**Comment by swissspidy** [comment]
> I dont see that in the code base.

@spacedmonkey it's in the PR description, not the code base.

**Comment by spacedmonkey** [comment]
> @spacedmonkey it's in the PR description, not the code base.
When I think todo, I think a code level comment. 

I personally don't think we should merge this as is, as it isn't really usable until the loading icon is not shown.

File: assets/src/story-embed-block/block/storyPlayer.js (hunk 15 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/storyPlayer.js:38]: Marking this here as resolved as this will be a non-issue once the player loads properly in the editor.

File: includes/Embed_Block.php [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+	/**
+	 * Renders the block type output for given attributes.
+	 *
+	 * @param array  $attributes Block attributes.
+	 * @param string $content    Block content.
+	 *
+	 * @return string Rendered block type output.
+	 */
+	public function render_block( array $attributes, $content ) {
+		// The only 2 mandatory attributes.
```
Comment (spacedmonkey) [inline includes/Embed_Block.php:148]: Should we have a different render for `is_feed()`? Maybe something that doesn't require js at all.

File: includes/REST_API/Embed_Controller.php (hunk 20 above)
Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:97]: I think we can break this off to another ticket.

File: includes/REST_API/Embed_Controller.php (hunk 17 above)
Comment (spacedmonkey) [reply includes/REST_API/Embed_Controller.php:190]: Okay, let's leave this for now.

File: includes/REST_API/Embed_Controller.php [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+
+		if ( ! $html ) {
+			set_transient( $cache_key, 'rest_invalid_story', $cache_ttl );
+			return new WP_Error( 'rest_invalid_story', __( 'URL is not a story', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$data = $this->get_data_from_document( $html );
+
+		if ( ! $data ) {
+			set_transient( $cache_key, 'rest_invalid_story', $cache_ttl );
```
Comment (spacedmonkey) [inline includes/REST_API/Embed_Controller.php]: Should we really be caching an error here? This is a bit of an anti pattern.

File: assets/src/story-embed-block/block/edit.js [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+        })
+          .then((data) => {
+            setCannotEmbed(!data?.title);
+            setStoryData(data);
+            setAttributes({
+              url: localURL,
+            });
+          })
+          .catch((data) => {
+            setStoryData(data);
```
Comment (spacedmonkey) [inline assets/src/story-embed-block/block/edit.js]: Why are you doing here?

File: assets/src/story-embed-block/block/edit.js [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+    }
+  }, [showLoadingIndicator, showPlaceholder]);
+
+  const fetchStoryData = useCallback(
+    (url) => {
+      if (!url) {
+        return;
+      }
+
+      try {
```
Comment (spacedmonkey) [inline assets/src/story-embed-block/block/edit.js:81]: Not sure I love mixing try / catch with then / catch. This make the code a little confusing. Can we just use async here and use await and a try / catch?

**Review by spacedmonkey — DISMISSED** [review]
I found a number of issues with the PR as it stands. I have listed below with screens. 

1. Add an embed story. Make the story bigger than the width of the content. Save post. Try and edit again. See video - https://youtu.be/RNA0pvaZwAg
^ This is a show stopped for merge. Can't edit the story after insert is an issue. My question is, why is the story have options to change the size in the first place? All other embed blocks are responsive. I think for biggest theme compatibility,  responsive is the way forward. 
2. Poster image is not previewed in doc sidebar. It is hard to know if need to replace the poster if you can't see it.
Screenshot 2020-05-18 at 11 15 38
3. The embed in the feed doesn't work at all and output invalid html. For feeds, I think we should have a simple embed, like an image with a hyperlink. 
Screenshot 2020-05-18 at 11 19 42
4. If you align the story right, if doesn't clear fix correctly. Not sure if this is a theme issue. 
Screenshot 2020-05-18 at 11 14 06
5. Preview in block list breaks out of preview area. ( This really is a nit pick ). 
Screenshot 2020-05-18 at 11 53 07
6. Not sure we can do anything about this, but if there is not enough space, then our block edit doesn't really work. 
Screenshot 2020-05-18 at 11 32 43

Otherwise, there are a couple of questions and nit picks, but otherwise we are looking good.

File: assets/src/story-embed-block/block/edit.js (hunk 24 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/edit.js]: In the case of an error, `data` contains an error object from the REST API that we can use to show an error to the user. `storyData?.message` is passed to `EmbedPlaceholder` because of this.

File: includes/REST_API/Embed_Controller.php (hunk 23 above)
Comment (swissspidy) [reply includes/REST_API/Embed_Controller.php]: No strong opinion here; removed caching errors now.

File: includes/Embed_Block.php (hunk 22 above)
Comment (swissspidy) [reply includes/Embed_Block.php:148]: Good idea. Implemented now.

File: assets/src/story-embed-block/block/edit.js (hunk 25 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/edit.js:81]: Refactored this now

File: assets/src/story-embed-block/block/embedControls.js [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+            onClick={switchBackToURLInput}
+          />
+        </ToolbarGroup>
+      </BlockControls>
+      <InspectorControls>
+        <PanelBody title={__('Embed Settings', 'web-stories')}>
+          <PanelRow>
+            <BaseControl>
+              <TextControl
+                type="string"
```
Comment (spacedmonkey) [inline assets/src/story-embed-block/block/embedControls.js]:
```suggestion
                type="text"
```

Comment (spacedmonkey) [reply assets/src/story-embed-block/block/embedControls.js]: This value is passed to the `<input` type and string is not valid.

**Comment by swissspidy** [comment]
> Add an embed story. Make the story bigger than the width of the content. Save post. Try and edit again.

Fixed. Max width has been reduced.

> Poster image is not previewed in doc sidebar. 

Fixed. Poster is now displayed in sidebar.

> The embed in the feed doesn't work at all and output invalid html. For feeds, I think we should have a simple embed, like an image with a hyperlink.

Fixed. An image with link is now shown in feeds, or just a link if there is no poster.

> If you align the story right, if doesn't clear fix correctly. Not sure if this is a theme issue.

Seems like a Gutenberg issue? e.g. https://github.com/WordPress/gutenberg/issues/10299

> Preview in block list breaks out of preview area. ( This really is a nit pick ).

Cannot reproduce this in current version of Gutenberg:

Screenshot 2020-05-18 at 13 53 21

> Not sure we can do anything about this, but if there is not enough space, then our block edit doesn't really work.

Not really. It's a common issue in Gutenberg. Example:

Screenshot 2020-05-18 at 14 27 43

File: assets/src/story-embed-block/block/embedControls.js (hunk 26 above)
Comment (swissspidy) [reply assets/src/story-embed-block/block/embedControls.js]: Good catch! Fixed in 8ab889c342e2f95ecbc0887a167e0a3dd5f16eb7

**Comment by spacedmonkey** [comment]
> Cannot reproduce this in current version of Gutenberg:
> 
> Screenshot 2020-05-18 at 13 53 21

I am not running the gutenberg plugin, I am just running whatever ships with 5.4.1

**Comment by spacedmonkey** [comment]
I still have issues with selecting the stories block. 

https://youtu.be/E8QJqDfyvS4

**Comment by swissspidy** [comment]
> I am not running the Gutenberg plugin, I am just running whatever ships with 5.4.1

They must have since improved the [`BlockPreview`](https://github.com/WordPress/gutenberg/tree/master/packages/block-editor/src/components/block-preview) component then 🤷 In WP 5.4.1 the preview only breaks out of the preview area because Gutenberg applies some margin there. Nothing we can really do something about.

> I still have issues with selecting the stories block.

Mind checking out the latest version? I tried adding an invisible overlay like Gutenberg does for its own embed blocks. This makes selection a bit better.

**Review by spacedmonkey — APPROVED** [review]
This is great work @swissspidy . I am very happy to approve it!
